File: study-6.js
// ==========================================================
// PENROSE PARTICLES
// STUDY 06
// CAN YOU TELL WHICH CUSTOMER IS FAKE?
//
// Website presentation layer.
// No JSON file or fetch() required.
// ==========================================================


// ==========================================================
// 1. STUDY RESULTS
// ==========================================================
//
// These values are displayed directly on the webpage.
//
// IMPORTANT:
// Classifier accuracy and synthetic-data similarity should
// ultimately be replaced with the results produced by the
// Python experiment.
// ==========================================================

const studyResults = {

    // ------------------------------------------------------
    // DATASET
    // ------------------------------------------------------

    realSessions: 12330,

    syntheticSessions: 12330,


    // ------------------------------------------------------
    // CLASSIFIER
    // ------------------------------------------------------
    //
    // Placeholder until the Python experiment is run.
    //
    // 50% = classifier is basically guessing
    // Higher = classifier can detect synthetic observations
    // ------------------------------------------------------

    detectionAccuracy: 50.0,

    rocAuc: 0.500,


    // ------------------------------------------------------
    // DISTRIBUTION SIMILARITY
    // ------------------------------------------------------
    //
    // Placeholder until calculated by Python.
    // ------------------------------------------------------

    averageSimilarity: 0,


    // ------------------------------------------------------
    // PURCHASE RATE
    // ------------------------------------------------------

    realPurchaseRate: 15.47,

    syntheticPurchaseRate: 0,


    // ------------------------------------------------------
    // RETURNING VISITORS
    // ------------------------------------------------------

    realReturningVisitors: 85.57,

    syntheticReturningVisitors: 0,


    // ------------------------------------------------------
    // WEEKEND SESSIONS
    // ------------------------------------------------------

    realWeekendSessions: 23.26,

    syntheticWeekendSessions: 0,


    // ------------------------------------------------------
    // VARIABLE SIMILARITY
    // ------------------------------------------------------
    //
    // These remain zero until we insert the actual results
    // from the Python synthetic-data experiment.
    // ------------------------------------------------------

    similarity: {

        Month: 0,

        VisitorType: 0,

        TrafficType: 0,

        Weekend: 0,

        PagesGroup: 0,

        DurationGroup: 0,

        BounceGroup: 0,

        ExitGroup: 0,

        ValueGroup: 0,

        Revenue: 0

    }

};


// ==========================================================
// 2. FRIENDLY VARIABLE NAMES
// ==========================================================

const friendlyNames = {

    Month:
        "Month",

    VisitorType:
        "Visitor type",

    TrafficType:
        "Traffic source",

    Weekend:
        "Weekend",

    PagesGroup:
        "Pages viewed",

    DurationGroup:
        "Session duration",

    BounceGroup:
        "Bounce behaviour",

    ExitGroup:
        "Exit behaviour",

    ValueGroup:
        "Page value",

    Revenue:
        "Purchase outcome"

};


// ==========================================================
// 3. FORMAT PERCENTAGES
// ==========================================================

function percent(value) {

    return `${Number(value).toFixed(1)}%`;

}


// ==========================================================
// 4. CALCULATE DIFFERENCE
// ==========================================================

function difference(real, synthetic) {

    const diff = Math.abs(
        Number(real) - Number(synthetic)
    );

    return `${diff.toFixed(1)} pp`;

}


// ==========================================================
// 5. CLASSIFIER INTERPRETATION
// ==========================================================

function classifierInterpretation(accuracy) {

    if (accuracy < 55) {

        return (
            "The classifier performed close to random guessing. " +
            "For the variables tested, the real and synthetic " +
            "shopping sessions were difficult to distinguish."
        );

    }


    if (accuracy < 70) {

        return (
            "The classifier detected some differences between " +
            "the real and synthetic shopping sessions, although " +
            "the two populations still shared substantial " +
            "statistical structure."
        );

    }


    return (
        "The classifier was able to distinguish the synthetic " +
        "sessions from the real sessions relatively well. " +
        "This suggests that important parts of the original " +
        "data distribution were not completely reproduced."
    );

}


// ==========================================================
// 6. BUILD SIMILARITY BARS
// ==========================================================

function buildSimilarityBars(variables) {

    const container =
        document.getElementById(
            "similarityList"
        );


    if (!container) {

        return;

    }


    container.innerHTML = "";


    // Convert object to array and sort
    // from highest similarity to lowest.

    const sortedVariables =
        Object.entries(variables)
            .sort(
                (a, b) => b[1] - a[1]
            );


    sortedVariables.forEach(
        ([variable, value]) => {


            // ----------------------------------------------
            // CONTAINER
            // ----------------------------------------------

            const item =
                document.createElement(
                    "div"
                );

            item.className =
                "similarity-item";


            // ----------------------------------------------
            // VARIABLE NAME
            // ----------------------------------------------

            const name =
                document.createElement(
                    "span"
                );

            name.className =
                "similarity-name";

            name.textContent =
                friendlyNames[variable]
                || variable;


            // ----------------------------------------------
            // BAR TRACK
            // ----------------------------------------------

            const track =
                document.createElement(
                    "div"
                );

            track.className =
                "similarity-track";


            // ----------------------------------------------
            // BAR
            // ----------------------------------------------

            const fill =
                document.createElement(
                    "div"
                );

            fill.className =
                "similarity-fill";


            track.appendChild(fill);


            // ----------------------------------------------
            // VALUE
            // ----------------------------------------------

            const number =
                document.createElement(
                    "span"
                );

            number.className =
                "similarity-value";

            number.textContent =
                percent(value);


            // ----------------------------------------------
            // ADD EVERYTHING
            // ----------------------------------------------

            item.appendChild(name);

            item.appendChild(track);

            item.appendChild(number);

            container.appendChild(item);


            // ----------------------------------------------
            // ANIMATE BAR
            // ----------------------------------------------

            requestAnimationFrame(() => {

                fill.style.width =
                    `${value}%`;

            });

        }
    );

}


// ==========================================================
// 7. DISPLAY TOP METRICS
// ==========================================================

function displayTopMetrics() {


    // ------------------------------------------------------
    // REAL SESSIONS
    // ------------------------------------------------------

    const realRows =
        document.getElementById(
            "realRows"
        );

    if (realRows) {

        realRows.textContent =
            studyResults
                .realSessions
                .toLocaleString();

    }


    // ------------------------------------------------------
    // SYNTHETIC SESSIONS
    // ------------------------------------------------------

    const syntheticRows =
        document.getElementById(
            "syntheticRows"
        );

    if (syntheticRows) {

        syntheticRows.textContent =
            studyResults
                .syntheticSessions
                .toLocaleString();

    }


    // ------------------------------------------------------
    // DETECTION ACCURACY
    // ------------------------------------------------------

    const accuracy =
        document.getElementById(
            "accuracy"
        );

    if (accuracy) {

        accuracy.textContent =
            percent(
                studyResults
                    .detectionAccuracy
            );

    }


    // ------------------------------------------------------
    // LARGE ACCURACY NUMBER
    // ------------------------------------------------------

    const accuracyLarge =
        document.getElementById(
            "accuracyLarge"
        );

    if (accuracyLarge) {

        accuracyLarge.textContent =
            percent(
                studyResults
                    .detectionAccuracy
            );

    }


    // ------------------------------------------------------
    // SIMILARITY
    // ------------------------------------------------------

    const similarity =
        document.getElementById(
            "similarity"
        );

    if (similarity) {

        similarity.textContent =
            percent(
                studyResults
                    .averageSimilarity
            );

    }

}


// ==========================================================
// 8. DISPLAY CLASSIFIER EXPLANATION
// ==========================================================

function displayClassifierResult() {

    const element =
        document.getElementById(
            "classifierInterpretation"
        );


    if (!element) {

        return;

    }


    element.textContent =
        classifierInterpretation(
            studyResults
                .detectionAccuracy
        );

}


// ==========================================================
// 9. DISPLAY PURCHASE RATE
// ==========================================================

function displayPurchaseRate() {


    const real =
        document.getElementById(
            "realPurchase"
        );


    const synthetic =
        document.getElementById(
            "syntheticPurchase"
        );


    const differenceElement =
        document.getElementById(
            "purchaseDifference"
        );


    if (real) {

        real.textContent =
            percent(
                studyResults
                    .realPurchaseRate
            );

    }


    if (synthetic) {

        synthetic.textContent =
            percent(
                studyResults
                    .syntheticPurchaseRate
            );

    }


    if (differenceElement) {

        differenceElement.textContent =
            difference(
                studyResults
                    .realPurchaseRate,

                studyResults
                    .syntheticPurchaseRate
            );

    }

}


// ==========================================================
// 10. DISPLAY RETURNING VISITORS
// ==========================================================

function displayReturningVisitors() {


    const real =
        document.getElementById(
            "realReturning"
        );


    const synthetic =
        document.getElementById(
            "syntheticReturning"
        );


    const differenceElement =
        document.getElementById(
            "returningDifference"
        );


    if (real) {

        real.textContent =
            percent(
                studyResults
                    .realReturningVisitors
            );

    }


    if (synthetic) {

        synthetic.textContent =
            percent(
                studyResults
                    .syntheticReturningVisitors
            );

    }


    if (differenceElement) {

        differenceElement.textContent =
            difference(
                studyResults
                    .realReturningVisitors,

                studyResults
                    .syntheticReturningVisitors
            );

    }

}


// ==========================================================
// 11. DISPLAY WEEKEND SESSIONS
// ==========================================================

function displayWeekendSessions() {


    const real =
        document.getElementById(
            "realWeekend"
        );


    const synthetic =
        document.getElementById(
            "syntheticWeekend"
        );


    const differenceElement =
        document.getElementById(
            "weekendDifference"
        );


    if (real) {

        real.textContent =
            percent(
                studyResults
                    .realWeekendSessions
            );

    }


    if (synthetic) {

        synthetic.textContent =
            percent(
                studyResults
                    .syntheticWeekendSessions
            );

    }


    if (differenceElement) {

        differenceElement.textContent =
            difference(
                studyResults
                    .realWeekendSessions,

                studyResults
                    .syntheticWeekendSessions
            );

    }

}


// ==========================================================
// 12. START STUDY
// ==========================================================

function initialiseStudy() {


    displayTopMetrics();


    displayClassifierResult();


    displayPurchaseRate();


    displayReturningVisitors();


    displayWeekendSessions();


    buildSimilarityBars(
        studyResults.similarity
    );

}


// ==========================================================
// 13. RUN WHEN PAGE LOADS
// ==========================================================

document.addEventListener(
    "DOMContentLoaded",
    initialiseStudy
);

File: study-6.py
# ============================================================
# PENROSE PARTICLES
# STUDY 6 — CAN YOU TELL WHICH CUSTOMER IS FAKE?
#
# Synthetic customer behaviour using a Bayesian Network
# ============================================================

import json
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, roc_auc_score


# ------------------------------------------------------------
# 1. SETTINGS
# ------------------------------------------------------------

RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)

DATA_FILE = "data/online_shoppers_intention.csv"
OUTPUT_FILE = "study-6-results.json"


# ------------------------------------------------------------
# 2. LOAD REAL DATA
# ------------------------------------------------------------

df = pd.read_csv(DATA_FILE)

print("Real dataset loaded.")
print("Rows:", len(df))
print("Columns:", len(df.columns))


# ------------------------------------------------------------
# 3. SELECT VARIABLES
# ------------------------------------------------------------
#
# We deliberately use a smaller group of variables so the
# resulting model remains understandable.
#
# ProductRelated          = number of product pages visited
# ProductRelated_Duration = time spent on product pages
# BounceRates             = session bounce rate
# ExitRates               = session exit rate
# PageValues              = estimated value of pages visited
# Month                   = month of session
# TrafficType             = traffic source category
# VisitorType             = returning/new visitor
# Weekend                 = whether visit occurred on weekend
# Revenue                 = whether session resulted in purchase
# ------------------------------------------------------------

columns = [
    "ProductRelated",
    "ProductRelated_Duration",
    "BounceRates",
    "ExitRates",
    "PageValues",
    "Month",
    "TrafficType",
    "VisitorType",
    "Weekend",
    "Revenue"
]

real = df[columns].copy()


# ------------------------------------------------------------
# 4. CLEAN DATA
# ------------------------------------------------------------

real["Weekend"] = real["Weekend"].astype(bool)
real["Revenue"] = real["Revenue"].astype(bool)

real = real.dropna().reset_index(drop=True)

print("Rows after cleaning:", len(real))


# ------------------------------------------------------------
# 5. CREATE DISCRETE BEHAVIOURAL STATES
# ------------------------------------------------------------
#
# Bayesian networks work naturally with conditional
# probabilities. Continuous variables are therefore grouped
# into behavioural categories.
# ------------------------------------------------------------

real["PagesGroup"] = pd.cut(
    real["ProductRelated"],
    bins=[-1, 5, 15, 40, np.inf],
    labels=["Very Low", "Low", "Medium", "High"]
)

real["DurationGroup"] = pd.cut(
    real["ProductRelated_Duration"],
    bins=[-1, 300, 1000, 3000, np.inf],
    labels=["Short", "Medium", "Long", "Very Long"]
)

real["BounceGroup"] = pd.cut(
    real["BounceRates"],
    bins=[-0.001, 0.01, 0.05, 0.10, np.inf],
    labels=["Very Low", "Low", "Medium", "High"]
)

real["ExitGroup"] = pd.cut(
    real["ExitRates"],
    bins=[-0.001, 0.02, 0.05, 0.10, np.inf],
    labels=["Very Low", "Low", "Medium", "High"]
)

real["ValueGroup"] = pd.cut(
    real["PageValues"],
    bins=[-0.001, 0.01, 5, 20, np.inf],
    labels=["None", "Low", "Medium", "High"]
)


# ------------------------------------------------------------
# 6. HELPER FUNCTION
# ------------------------------------------------------------

def sample_from_distribution(series):
    """
    Sample one value according to the observed probability
    distribution in a pandas Series.
    """

    probabilities = series.value_counts(normalize=True)

    return np.random.choice(
        probabilities.index,
        p=probabilities.values
    )


# ------------------------------------------------------------
# 7. GENERATE SYNTHETIC CUSTOMERS
# ------------------------------------------------------------
#
# Simplified Bayesian dependency structure:
#
# Month ───────────────┐
#                     ↓
# VisitorType → PagesGroup
#                     ↓
#               DurationGroup
#                     ↓
#                 ExitGroup
#                     ↓
#                 Revenue
#
# TrafficType ─────────┘
#
# Revenue also depends on engagement/value information.
#
# IMPORTANT:
# These arrows represent probabilistic dependencies used for
# simulation. They should not automatically be interpreted
# as proven causal relationships.
# ------------------------------------------------------------

synthetic_rows = []

n_synthetic = len(real)

for i in range(n_synthetic):

    # Root variables
    month = sample_from_distribution(real["Month"])
    visitor = sample_from_distribution(real["VisitorType"])
    traffic = sample_from_distribution(real["TrafficType"])
    weekend = sample_from_distribution(real["Weekend"])

    # --------------------------------------------------------
    # Pages visited conditional on visitor type
    # --------------------------------------------------------

    subset = real[
        real["VisitorType"] == visitor
    ]

    pages_group = sample_from_distribution(
        subset["PagesGroup"]
    )

    # --------------------------------------------------------
    # Duration conditional on page activity
    # --------------------------------------------------------

    subset = real[
        real["PagesGroup"] == pages_group
    ]

    duration_group = sample_from_distribution(
        subset["DurationGroup"]
    )

    # --------------------------------------------------------
    # Bounce behaviour conditional on engagement
    # --------------------------------------------------------

    subset = real[
        (real["PagesGroup"] == pages_group) &
        (real["DurationGroup"] == duration_group)
    ]

    if len(subset) < 10:
        subset = real

    bounce_group = sample_from_distribution(
        subset["BounceGroup"]
    )

    # --------------------------------------------------------
    # Exit behaviour conditional on bounce behaviour
    # --------------------------------------------------------

    subset = real[
        real["BounceGroup"] == bounce_group
    ]

    exit_group = sample_from_distribution(
        subset["ExitGroup"]
    )

    # --------------------------------------------------------
    # Page value conditional on engagement
    # --------------------------------------------------------

    subset = real[
        (real["PagesGroup"] == pages_group) &
        (real["DurationGroup"] == duration_group)
    ]

    if len(subset) < 10:
        subset = real

    value_group = sample_from_distribution(
        subset["ValueGroup"]
    )

    # --------------------------------------------------------
    # Purchase conditional on several behavioural variables
    # --------------------------------------------------------

    subset = real[
        (real["VisitorType"] == visitor) &
        (real["PagesGroup"] == pages_group) &
        (real["ValueGroup"] == value_group)
    ]

    if len(subset) < 20:
        subset = real[
            (real["PagesGroup"] == pages_group) &
            (real["ValueGroup"] == value_group)
        ]

    if len(subset) < 20:
        subset = real

    revenue = sample_from_distribution(
        subset["Revenue"]
    )

    synthetic_rows.append({
        "Month": month,
        "VisitorType": visitor,
        "TrafficType": traffic,
        "Weekend": weekend,
        "PagesGroup": pages_group,
        "DurationGroup": duration_group,
        "BounceGroup": bounce_group,
        "ExitGroup": exit_group,
        "ValueGroup": value_group,
        "Revenue": revenue
    })


synthetic = pd.DataFrame(synthetic_rows)

print("Synthetic customers generated:", len(synthetic))


# ------------------------------------------------------------
# 8. PREPARE REAL DATA FOR COMPARISON
# ------------------------------------------------------------

comparison_columns = [
    "Month",
    "VisitorType",
    "TrafficType",
    "Weekend",
    "PagesGroup",
    "DurationGroup",
    "BounceGroup",
    "ExitGroup",
    "ValueGroup",
    "Revenue"
]

real_model = real[comparison_columns].copy()

real_model["Dataset"] = 0
synthetic["Dataset"] = 1


# ------------------------------------------------------------
# 9. MIX REAL + SYNTHETIC DATA
# ------------------------------------------------------------

combined = pd.concat(
    [real_model, synthetic],
    ignore_index=True
)

X = combined.drop(columns=["Dataset"])
y = combined["Dataset"]


# ------------------------------------------------------------
# 10. CLASSIFIER
# ------------------------------------------------------------
#
# The classifier's job is unusual:
#
# 0 = REAL
# 1 = SYNTHETIC
#
# If the datasets are very similar, distinguishing them should
# be difficult.
# ------------------------------------------------------------

categorical_features = list(X.columns)

preprocessor = ColumnTransformer(
    transformers=[
        (
            "categorical",
            OneHotEncoder(
                handle_unknown="ignore"
            ),
            categorical_features
        )
    ]
)

classifier = RandomForestClassifier(
    n_estimators=300,
    max_depth=8,
    random_state=RANDOM_STATE,
    n_jobs=-1
)

model = Pipeline([
    ("preprocessor", preprocessor),
    ("classifier", classifier)
])


# ------------------------------------------------------------
# 11. TRAIN / TEST SPLIT
# ------------------------------------------------------------

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.30,
    random_state=RANDOM_STATE,
    stratify=y
)

model.fit(X_train, y_train)


# ------------------------------------------------------------
# 12. TEST CLASSIFIER
# ------------------------------------------------------------

predictions = model.predict(X_test)

probabilities = model.predict_proba(X_test)[:, 1]

accuracy = accuracy_score(
    y_test,
    predictions
)

auc = roc_auc_score(
    y_test,
    probabilities
)

print()
print("REAL VS SYNTHETIC CLASSIFIER")
print("----------------------------")
print("Accuracy:", round(accuracy, 4))
print("ROC AUC:", round(auc, 4))


# ------------------------------------------------------------
# 13. PURCHASE RATE COMPARISON
# ------------------------------------------------------------

real_purchase_rate = (
    real_model["Revenue"].mean() * 100
)

synthetic_purchase_rate = (
    synthetic["Revenue"].mean() * 100
)


# ------------------------------------------------------------
# 14. RETURNING VISITOR COMPARISON
# ------------------------------------------------------------

real_returning = (
    (real_model["VisitorType"] == "Returning_Visitor")
    .mean() * 100
)

synthetic_returning = (
    (synthetic["VisitorType"] == "Returning_Visitor")
    .mean() * 100
)


# ------------------------------------------------------------
# 15. WEEKEND COMPARISON
# ------------------------------------------------------------

real_weekend = (
    real_model["Weekend"].mean() * 100
)

synthetic_weekend = (
    synthetic["Weekend"].mean() * 100
)


# ------------------------------------------------------------
# 16. DISTRIBUTION SIMILARITY
# ------------------------------------------------------------

def distribution_distance(real_series, synthetic_series):
    """
    Total Variation Distance.

    0 means distributions are identical.
    1 means completely different.
    """

    real_dist = real_series.value_counts(normalize=True)
    synth_dist = synthetic_series.value_counts(normalize=True)

    categories = set(real_dist.index).union(
        set(synth_dist.index)
    )

    distance = 0

    for category in categories:

        real_probability = real_dist.get(category, 0)
        synth_probability = synth_dist.get(category, 0)

        distance += abs(
            real_probability -
            synth_probability
        )

    return distance / 2


variables_to_compare = [
    "Month",
    "VisitorType",
    "TrafficType",
    "Weekend",
    "PagesGroup",
    "DurationGroup",
    "BounceGroup",
    "ExitGroup",
    "ValueGroup",
    "Revenue"
]

similarity_results = {}

for variable in variables_to_compare:

    distance = distribution_distance(
        real_model[variable],
        synthetic[variable]
    )

    similarity = (1 - distance) * 100

    similarity_results[variable] = round(
        similarity,
        2
    )


average_similarity = np.mean(
    list(similarity_results.values())
)


# ------------------------------------------------------------
# 17. PURCHASE BY VISITOR TYPE
# ------------------------------------------------------------

real_purchase_by_visitor = (
    real_model
    .groupby("VisitorType", observed=False)["Revenue"]
    .mean()
    .mul(100)
    .round(2)
    .to_dict()
)

synthetic_purchase_by_visitor = (
    synthetic
    .groupby("VisitorType", observed=False)["Revenue"]
    .mean()
    .mul(100)
    .round(2)
    .to_dict()
)


# ------------------------------------------------------------
# 18. FEATURE IMPORTANCE
# ------------------------------------------------------------

encoder = (
    model
    .named_steps["preprocessor"]
    .named_transformers_["categorical"]
)

feature_names = encoder.get_feature_names_out(
    categorical_features
)

importances = (
    model
    .named_steps["classifier"]
    .feature_importances_
)

importance_df = pd.DataFrame({
    "feature": feature_names,
    "importance": importances
})

importance_df = (
    importance_df
    .sort_values(
        "importance",
        ascending=False
    )
    .head(8)
)

feature_importance = []

for _, row in importance_df.iterrows():

    feature_importance.append({
        "feature": row["feature"],
        "importance": round(
            float(row["importance"] * 100),
            2
        )
    })


# ------------------------------------------------------------
# 19. INTERPRET CLASSIFIER RESULT
# ------------------------------------------------------------

if accuracy < 0.55:

    classifier_interpretation = (
        "The classifier performed close to random guessing. "
        "For the variables tested, real and synthetic sessions "
        "were difficult to distinguish."
    )

elif accuracy < 0.70:

    classifier_interpretation = (
        "The classifier detected some differences between the "
        "real and synthetic sessions, although the two datasets "
        "still shared substantial statistical structure."
    )

else:

    classifier_interpretation = (
        "The classifier could distinguish the synthetic sessions "
        "from real sessions relatively well. This suggests that "
        "important parts of the real data distribution were not "
        "fully reproduced."
    )


# ------------------------------------------------------------
# 20. EXPORT RESULTS FOR WEBSITE
# ------------------------------------------------------------

results = {

    "study": {
        "title": "Can You Tell Which Customer Is Fake?",
        "real_rows": int(len(real_model)),
        "synthetic_rows": int(len(synthetic)),
        "random_state": RANDOM_STATE
    },

    "classifier": {
        "accuracy": round(
            float(accuracy * 100),
            2
        ),
        "roc_auc": round(
            float(auc),
            3
        ),
        "interpretation": classifier_interpretation
    },

    "similarity": {
        "average": round(
            float(average_similarity),
            2
        ),
        "variables": similarity_results
    },

    "purchase": {
        "real": round(
            float(real_purchase_rate),
            2
        ),
        "synthetic": round(
            float(synthetic_purchase_rate),
            2
        )
    },

    "returning_visitors": {
        "real": round(
            float(real_returning),
            2
        ),
        "synthetic": round(
            float(synthetic_returning),
            2
        )
    },

    "weekend": {
        "real": round(
            float(real_weekend),
            2
        ),
        "synthetic": round(
            float(synthetic_weekend),
            2
        )
    },

    "purchase_by_visitor": {
        "real": real_purchase_by_visitor,
        "synthetic": synthetic_purchase_by_visitor
    },

    "feature_importance": feature_importance
}


with open(
    OUTPUT_FILE,
    "w",
    encoding="utf-8"
) as file:

    json.dump(
        results,
        file,
        indent=4
    )


print()
print("Study complete.")
print("Results exported to:", OUTPUT_FILE)